### PR TITLE
Add a Rust implementation

### LINKED ISSUE #289: Make a rust implementation

### PR DESCRIPTION
**Pull request by djmitche**
This is largely the work of @owlishDeveloper and @alopezz, as shown in the individual commits.

# Checklist

Before submitting a pull request, please check the following:

* [x] All tests pass for you on your machine for all implementations (all implementations must behave identically)
* [x] New tests have been added for any new functionality or to prevent regressions of a bugfix
* [x] Added a short description of the change to `newsfragments/$issue.bugfix` (for fixes) or `.feature` (for new features) or `.doc` (for documentation-only changes)

Fixes #289.

### CODE REVIEW

File: specification.yml [hunk 1]
```diff
@@ -32,6 +32,11 @@ context:  {}
 template: {$fromNow: '2 days 3 hours'}
 result:   '2017-01-21T19:27:20.974Z'
 ---
+title:    $eval without context
```
Comment (imbstack) [inline specification.yml:35]: surprised we didn't have this before now!

File: specification.yml [hunk 2]
```diff
@@ -1531,40 +1509,20 @@ template: {$eval: 'floor(key)'}
...
-template: {$eval: 'floor(2.003)'}
-result: 2.0
----
-title: floor (5)
+title: floor (3)
 context: {key: 1}
 template: {$eval: 'floor(key)'}
 result: 1
 ---
-title: floor (6)
```
Comment (imbstack) [inline specification.yml:1553]: why all of these specs removed?

**Review by imbstack — COMMENTED** [review]
I'm still reviewing most of this but wanted to ask now about the removals of a lot of specifications?

File: specification.yml (hunk 2 above)
Comment (alopezz) [reply specification.yml:1553]: I discussed this with Dustin and none of us could figure out what these tests were actually testing. In his words:

> so in combination that's testing that the JSON library considers 3 and 3.0 the same.

Comment (djmitche) [reply specification.yml:1553]: "combination" there meaning, the deleted test case in combination with the previous test case.  The inputs are the same, so it's just checking the equality operator on JSON objects (which in this case is `serde_json`, and in this case does not consider 3 to equal 3.0, but that's not our problem to solve).

File: src/value.rs [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+            (Function::WithContext(f1), Function::WithContext(f2)) => {
+                *f1 as *const () == *f2 as *const ()
+            }
+            _ => false,
+        }
+    }
+}
+
+impl Function {
+    // TODO: pub(crate) constructor, take name for Debug
```
Comment (imbstack) [inline src/value.rs:39]: should this TODO be done before landing or converted into an issue? linking it here if the latter

**Review by imbstack — APPROVED** [review]
Very impressive!

File: src/value.rs (hunk 3 above)
Comment (djmitche) [reply src/value.rs]: I just fixed it .. I'll merge if it's green!

**Pull request merged**